### PR TITLE
Coral.

### PR DESCRIPTION
**Pull request by sofar**
Schematic based coral. Coral is fairly rare - you'll only find it in
the "hot" biomes, underwater. It has a dead coral stem and sometimes
dead ends. It's fairly blob-ish.

Harvesting coral kills it. If you attempt to place live coral outside
of water, or if you drain the water from the coral it will quickly
die off.

This isn't finished per se. We can add variants for the dead coral
block for slabs and stairs, and possibly other uses. Another possible
thing to add may be to make coral actually grow according to coral
growth rules: up and sideways to y = -2 under proper circumstances,
although some growth limit should be made so it doesn't fill out the
entire sea eventually.

### CODE REVIEW

**Comment by sofar** [comment]
The schematic is fairly crude. Making it 2 high may be better. We can also remove the dead parts if it doesn't look good, just saying - it needs work.

**Comment by tobyplowy** [comment]
@sofar it's a good start but it's still a little bit empty, still really appreciate you do this 👍

**Comment by sofar** [comment]
The noise params were chosen to make it rare. The last thing I want to do is go from "the sea is empty" to "the sea is full". Small steps... small steps.

**Comment by tobyplowy** [comment]
@sofar but if they only spawn in hot water biomes what will happen with all the other ocean biomes for example cold water?

**Comment by sofar** [comment]
https://gist.github.com/sofar/92af687c856b6ae4688cf04564e923b1

schematic currently used.

**Comment by sofar** [comment]
@tobyplowy cold oceans are hostile to coral: it won't spawn there.

**Comment by tobyplowy** [comment]
@sofar And normal biomes so not hot but not cold either? Sorry that I'm being annoying I'm just so excited :)

**Comment by paramat** [comment]
Sofar so good. Maybe somehow we can add a second live coral colour. I'll experiment with noise params, how about widely seperated but dense patches of schematics?

tobyplowy research coral habitat.

**Comment by paramat** [comment]
> Most coral reefs exist in shallow waters less than 50 m deep

I tend to work on Minetest being 1:10 scale of realistic, due to cloud height, so we can keep these in very shallow water but of course always below surface to avoid coral death.

**Comment by C1ffisme** [comment]
Will coral have any gameplay use at some point?

**Comment by rubenwardy** [comment]
IRL coral can be used for jewelry and medicine. But I can't see either of these in MT, so it'll probably just be decorative

**Comment by DonBatman** [comment]
I like the idea but not fond of the color. Would be nice to see it more colorful. The texture might have to be a bit exaggerated to account for the darkening in water.
I would also like to see more colors.

**Comment by Fixer-007** [comment]
This texture has a lot of potential for me, at least as couches and beds...

**Comment by pithydon** [comment]
Interesting texture choice for the dead coral.

**Comment by tobyplowy** [comment]
@DonBatman more colors would be amazing thats true

**Comment by paramat** [comment]
Yeah just a start we'll probably add at least one extra colour.

**Comment by tobyplowy** [comment]
I'm so excited for this :smile:

**Comment by tobyplowy** [comment]
@sofar if the corals die when you harvest them how are you going to use them to make aquariums?

**Comment by sofar** [comment]
@tobyplowy Some things in nature are better left untouched and will never look as pretty.

(I'm not entirely opposed to making coral behave more like normal rocks, but I think it's an excellent opportunity to add an element that displays the importance environmental conservation in the game. The fact that it has a cost to gameplay is crucial to that)

**Comment by tobyplowy** [comment]
@sofar okay but It limits creativity :/

**Comment by paramat** [comment]
I support the dying feature, they're still pretty when white.

**Comment by sofar** [comment]
I have no issues with making these dyable/paintable.

**Comment by yyt16384** [comment]
@tobyplowy you can always use pistons :)

**Comment by paramat** [comment]
To be clear i meant dying not dyeing.

**Comment by tobyplowy** [comment]
@paramat painting them seems like an amazing idea (fix for them creativity problem)

**Comment by Fixer-007** [comment]
Slabs and stairs are needed for this.

**Comment by paramat** [comment]
Painting corals seems highly distasteful, that should be done by mod.

**Comment by tobyplowy** [comment]
@Fixer-007 for Coral or processed Coral?

**Comment by t4im** [comment]
That pun went rogue :smiley:
Great addition, those corals, sofar.

> how are you going to use them to make aquariums?

@tobyplowy, you can still build glass around them, then drain the sea. Some work, but you might be the only one with a coral aquarium then ;-)

**Comment by tobyplowy** [comment]
@sofar seems like a lot of people really like this XD

**Comment by tobyplowy** [comment]
@sofar are you still working on this?

**Comment by paramat** [comment]
All we need is another colour, then i need to tune the use of noise.

**Comment by tobyplowy** [comment]
@paramat @pithydon a light gentle pink color would be nice

**Comment by paramat** [comment]
This is actually mergeable, then i can tune the noise and another colour can be added later. Will test this.

**Comment by tobyplowy** [comment]
@paramat will celp also get added later?

**Comment by paramat** [comment]
Actually, this should wait until we have extra colours in the schematic, so that generated corals are colourful from the beginning. But this is nearly ready.

**Comment by DonBatman** [comment]
Any chance these models could be used?

https://forum.minetest.net/viewtopic.php?f=9&t=13425&sid=ac3b6796f34f91e9bd702a532582660c

**Comment by paramat** [comment]
They're beautiful, but will probably cause air bubbles, this is why we're using cubic corals.

**Comment by tobyplowy** [comment]
@DonBatman they look really nice but I don't think they fit the style of minetest (the models Not the textures) apart from the 3rd and the 4th one (the Grass like models) do fit the style

**Comment by tobyplowy** [comment]
@paramat oh man I can't wait for this to be done :D also is there a way to fix that bubble problem?

**Comment by paramat** [comment]
Yes but it's a complex fix, see the sea mods.

**Comment by Ekdohibs** [comment]
Well, it could be probably be fixed by modifying the model, although I'm not sure it would really work well.
However, it *is* a hack, so I think it would be good if support for nodes that can get flooded with water or other liquids was added to the engine.

**Comment by tenplus1** [comment]
+1 for "liquidplantlike" style.

**Comment by rubenwardy** [comment]
I welcome the effort to make the underwater more interesting. Although I feel like just adding a decorative node is a little boring - is there any way to add a use to this? It would be nice, as part of player progression, to require them to explore the depths (although MT oceans aren't that deep)

**Comment by tobyplowy** [comment]
@rubenwardy you have a good point there, also I really think oceans should go a little bit deeper @rubenwardy do you have any ideas?

**Comment by DonBatman** [comment]
@paramat any ideas about deeper oceans?

**Comment by paramat** [comment]
Mgv6 can't be changed. Mgv7 is semi-stable and is intended to be faithful to hmmmm's shallow-sea design. Mgv5 ocean can occasionally be as deep as 160 nodes in places.

**Comment by tobyplowy** [comment]
@paramat why do we actually have different versions of world generators?

**Comment by tobyplowy** [comment]
@paramat now that were talking about this this, is a little bit off topic but it's a pretty good feature so I'm going to ask it ( I don't know if you play Minecraft but since 1.7 i think, they added a complicated world generator you can actually change all the sliders to say maximum height of mountains and water level the density of trees and stuff like that) I think minetest really needs this.

**Comment by paramat** [comment]
Note i'm happy with adding corals even if there is no use other than decoration. It's another exotic ingredient for mod recipes to use (calcium carbonate).

**Comment by paramat** [comment]
I'm thinking add one more colour, the orange one from https://github.com/minetest/minetest_game/issues/1129#issuecomment-227205165 and call the 3 nodes coral_brown / orange / skeleton?
A manual-random distribution of colours will be needed in the schematic, then set random rotation for the schematic to compenste for this.

**Comment by tobyplowy** [comment]
@paramat will kelp get added when the bubble bug get fixed, or will that not happen? I'm just wondering :)

File: mods/default/nodes.lua [hunk 1]
```diff
@@ -415,6 +418,21 @@ minetest.register_node("default:clay", {
 	sounds = default.node_sound_dirt_defaults(),
 })
 
+minetest.register_node("default:coral_live", {
+	description = "Coral",
+	tiles = {"default_coral_live.png"},
+	groups = {cracky = 3},
+	drop = 'default:coral',
```
Comment (rubenwardy) [inline mods/default/nodes.lua]: inconsistent quote marks

**Comment by rubenwardy** [comment]
All the code looks good, although I can't comment on the decoration code as I haven't learned how to use them yet. This adds nice variety to underwater, and could be used for construction, artificial fish tanks, medicine, and jewellery

File: mods/default/nodes.lua (hunk 1 above)
Comment (paramat) [inline mods/default/nodes.lua]: I agree, for consistency. Both work but i've gradually been changing everything to " in game.

**Comment by paramat** [comment]
Note i'm happy to tune the noise parameters for the distribution, either when this is almost ready or in a following commit.

**Comment by rubenwardy** [comment]
I wish that Minetest supported underwater plants. Would love to see seaweed

**Comment by tobyplowy** [comment]
@rubenwardy me too :(

**Comment by Ferk** [comment]
I was actually checking on the code that draws the liquid faces, and it seems that there's some [legacy code that isn't really used now](https://github.com/minetest/minetest/blob/master/src/content_mapblock.cpp#L216) for rendering them that would have allowed things like 1 node tall seaweeds without creating bubbles.

However, when I enabled it ([by changing this](https://github.com/minetest/minetest/blob/stable-0.4/src/mapblock_mesh.cpp#L649)), it also brought along a lot of glitches, places where the faces were not rendered. I assume the changes were done to prevent that.

It would be very complex to hide the water faces when next to specific nodes without messing up.

**Comment by Megaf** [comment]
I don't like it the way it is right now. Maybe if it was shaped like coral and if there was some use added to it.

**Comment by everamzah** [comment]
Shape it like cactus.

**Comment by tobyplowy** [comment]
@everamzah I don't think It should be a cantus because that Would ruin it as a building block (also it's not possible until the bubble problem is fixed)

**Comment by everamzah** [comment]
Yeah, you're probably right.  Hopefully in-water nodes can be defined someday.  I mean, normal null space in a node is filled with what appears to be air, so it should be possible to fill it with what appears to be water.

**Comment by paramat** [comment]
Nodes added in #1324

**Comment by paramat** [comment]
#1324 merged. Will be completing the mapgen stuff later.

**Pull request closed without merging**